src/module.c

```diff
@@ -485,71 +485,96 @@ void handleCompaction(RedisModuleCtx *ctx, CompactionRule *rule, api_timestamp_t
+	  RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+
+	    double value;
+	    api_timestamp_t timestamp;
+	    if ((RedisModule_StringToDouble(valueStr, &value) != REDISMODULE_OK))
+	        return RedisModule_ReplyWithError(ctx, "TSDB: invalid value");
+
+	    if ((RedisModule_StringToLongLong(timestampStr, (long long int *) &timestamp) != REDISMODULE_OK)) {
+	        // if timestamp is "*", take current time (automatic timestamp)
+	        if(RMUtil_StringEqualsC(timestampStr, "*"))
+	            timestamp = (u_int64_t) time(NULL);
+	        else
+	            return RedisModule_ReplyWithError(ctx, "TSDB: invalid timestamp");
+	    }
+
+	    Series *series = NULL;
+
+	    if (argv != NULL && RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+	        // the key doesn't exist, lets check we have enough information to create one
+	        long long retentionSecs;
+	        long long maxSamplesPerChunk;
+	        size_t labelsCount;
+	        Label *labels;
+	        if (parseCreateArgs(ctx, argv, argc, &retentionSecs, &maxSamplesPerChunk, &labelsCount, &labels) != REDISMODULE_OK) {
+	            return REDISMODULE_ERR;
+	        }
+
+	        CreateTsKey(ctx, keyName, labels, labelsCount, retentionSecs, maxSamplesPerChunk, &series, &key);
+	        SeriesCreateRulesFromGlobalConfig(ctx, keyName, series, labels, labelsCount);
+	    } else if (RedisModule_ModuleTypeGetType(key) != SeriesType){
+	        return RedisModule_ReplyWithError(ctx, "TSDB: the key is not a TSDB key");
+	    } else {
+	        series = RedisModule_ModuleTypeGetValue(key);
+	    }
+
+	    int retval = SeriesAddSample(series, timestamp, value);
+	    int result = 0;
+	    if (retval == TSDB_ERR_TIMESTAMP_TOO_OLD) {
+	        RedisModule_ReplyWithError(ctx, "TSDB: timestamp is too old");
+	        result = REDISMODULE_ERR;
+	    } else if (retval != TSDB_OK) {
+	        RedisModule_ReplyWithError(ctx, "TSDB: Unknown Error");
+	        result = REDISMODULE_ERR;
+	    } else {
+	        // handle compaction rules
+	        CompactionRule *rule = series->rules;
+	        while (rule != NULL) {
+	            handleCompaction(ctx, rule, timestamp, value);
+	            rule = rule->nextRule;
+	        }
+	        RedisModule_ReplyWithLongLong(ctx, timestamp);
+	        result = REDISMODULE_OK;
+	    }
+	    RedisModule_CloseKey(key);
+	    return result;
+}
+
+
+int TSDB_madd(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+	   RedisModule_AutoMemory(ctx);
+
+	   if (argc < 4 || (argc-1)%3 != 0) {
+		   return RedisModule_WrongArity(ctx);
+	   }
+
+	   RedisModule_ReplyWithArray(ctx, (argc-1)/3);
+	   for(int i=1; i<argc ; i+=3){
+			RedisModuleString *keyName = argv[i];
+			RedisModuleString *timestampStr = argv[i+1];
+			RedisModuleString *valueStr = argv[i+2];
+			add(ctx, keyName, timestampStr, valueStr, NULL, -1);
```

Comment:
@gkorland what happens if some of the adds fail and the other dont??

Reply:
you get an array of results and errors see test_partial_madd